leasing/report/invoice/invoicing_review.py

```diff
@@ -458,108 +456,123 @@ def get_gaps_in_billing_periods_data(
-                l.id AS "lease_id",
-                l.start_date,
-                l.end_date,
-                r.id AS "rent_id",
-                r.start_date AS "rent_start_date",
-                r.end_date AS "rent_end_date",
-                i.id AS "invoice_id",
-                i.billing_period_start_date,
-                i.billing_period_end_date
-            FROM leasing_lease l
-            INNER JOIN leasing_leaseidentifier li
-                ON l.identifier_id = li.id
-            INNER JOIN leasing_rent r
-                ON l.id = r.lease_id
-            LEFT JOIN leasing_invoice i
-                ON l.id = i.lease_id
-            INNER JOIN leasing_invoicerow ir
-                ON i.id = ir.invoice_id
-            WHERE (l.end_date IS NULL OR l.end_date >= %(today)s)
-                AND (l.start_date IS NOT NULL AND l.start_date <= %(today)s)
-                AND l.service_unit_id = ANY(%(service_units)s)
-                AND l.deleted IS NULL
-                AND l.state NOT IN ('reservation', 'power_of_attorney')
-                AND r.deleted IS NULL
-                AND r."type" != 'free'
-                AND (r.start_date IS NULL OR r.start_date <= %(today)s)
-                AND (r.end_date IS NULL OR r.end_date >= %(today)s)
-                AND i.deleted IS NULL
-                AND i.deleted_by_cascade IS false
-                AND ir.deleted IS NULL
-                AND ir.receivable_type_id != ANY(%(excluded_receivable_type_ids)s)
-            ORDER BY li.identifier ASC;
+WITH invoices_sorted AS (
+    SELECT
+        i.id,
+        i.billing_period_start_date,
+        i.billing_period_end_date,
+        r.id AS rent_id,
+        l.id AS lease_id,
+        li.identifier AS lease_identifier,
+        COALESCE(c.name, (c.first_name || ' ' || c.last_name)) AS recipient_name,
+        COALESCE(r.start_date, l.start_date) AS rent_start_date,
+        COALESCE(r.end_date, l.end_date) AS rent_end_date,
+        LEAD(i.billing_period_start_date) OVER (PARTITION BY r.id ORDER BY i.billing_period_start_date)
+            AS next_start_date
+    FROM
+        leasing_invoice i
+    INNER JOIN
+        leasing_invoicerow ir ON ir.invoice_id = i.id
+    LEFT JOIN
+        leasing_rent r ON i.lease_id = r.lease_id
+    INNER JOIN
+        leasing_lease l ON r.lease_id = l.id
+    INNER JOIN
+        leasing_leaseidentifier li ON l.identifier_id = li.id
+    LEFT JOIN
+        leasing_receivabletype rt ON ir.receivable_type_id = rt.id
+    LEFT JOIN
+        leasing_contact c ON i.recipient_id = c.id
+    WHERE
+        ir.deleted IS NULL
+        AND i.deleted IS NULL
+        AND i.type IN ('charge')
+        AND r.deleted IS NULL
+        AND r."type" NOT IN ('free', 'manual')
+        AND rt.name NOT IN (
+            'Korko',
+            'Yhteismarkkinointi (sis. ALV)',
+            'Kiinteistötoimitukset (tonttijaot, lohkomiset, rekisteröimiskustannukset, rasitteet)',
+            'Rahavakuus'
```

Comment:
This will break if the names change in future, but I don't see any other good way to make this check without hardcoding ID's or similar.

Reply:
There is a test actually that ensures these exist: `test_excluded_receivable_type_ids()`